cpp/src/join/quadtree_point_to_nearest_polyline.cu

```diff
@@ -146,74 +163,178 @@ struct compute_quadtree_point_to_nearest_polyline {
+      // ```
+      //
+      // However, the above psuedocode produces values in an order such that the distance
+      // from a point to each polyline cannot be reduced with `thrust::reduce_by_key`:
+      // ```
+      //   point | polyline | distance
+      //       0 |        0 |     10.0
+      //       1 |        0 |     30.0
+      //       2 |        0 |     20.0
+      //       0 |        1 |     30.0
+      //       1 |        1 |     20.0
+      //       2 |        1 |     10.0
+      // ```
+      //
+      // In order to use `thrust::reduce_by_key` to compute the minimum distance from a point to
+      // the polylines in its quadrant, the above table needs to be sorted by `point` instead of
+      // `polyline`:
+      // ```
+      //   point | polyline | distance
+      //       0 |        0 |     10.0
+      //       0 |        1 |     30.0
+      //       1 |        0 |     30.0
+      //       1 |        1 |     20.0
+      //       2 |        0 |     20.0
+      //       2 |        1 |     10.0
+      // ```
+      //
+      // A naive approach would be to allocate memory for the above three columns, sort the
+      // columns by `point`, then use `thrust::reduce_by_key` to compute the min distances.
+      //
+      // The sizes of the intermediate buffers required can easily grow beyond available
+      // device memory, so a better approach is to use a Thrust iterator to yield values
+      // in the sorted order on demand instead, which is what we're doing here.
+      //
+
+      auto all_point_poly_indices_and_distances = thrust::make_transform_iterator(
+        thrust::make_counting_iterator(0u),
+        compute_point_poly_indices_and_distances<T,
+                                                 decltype(point_xys_iter),
+                                                 decltype(quad_offsets),
+                                                 decltype(quad_lengths)>{
+          point_xys_iter,
+          local_point_offsets.begin(),
+          local_point_offsets.end(),
+          quad_offsets,
+          quad_offsets + num_poly_quad_pairs,
+          quad_lengths,
+          poly_indices,
+          *cudf::column_device_view::create(poly_offsets, stream),
+          *cudf::column_device_view::create(poly_points_x, stream),
+          *cudf::column_device_view::create(poly_points_y, stream)});
+
+      auto all_point_indices =
+        thrust::make_transform_iterator(all_point_poly_indices_and_distances,
+                                        [] __device__(auto const &x) { return thrust::get<0>(x); });
+
+      // Allocate vectors for the distances min reduction
+      rmm::device_uvector<uint32_t> point_idxs(point_x.size(), stream);
+      rmm::device_uvector<uint32_t> poly_idxs(point_x.size(), stream);
+      rmm::device_uvector<T> distances(point_x.size(), stream);
+
+      // Fill distances with 0
+      thrust::fill(rmm::exec_policy(stream), distances.begin(), distances.end(), T{0});
```

Comment:
First, does this  really need to be filled with zero? Doesn't the reduce_by_key write to every element of the output values? 

If it does need to be initialized, like all Thrust APIs that are not in the `thrust::async` namespace, this will synchronize. We have a utility in libcudf, but  it is a detail API... So here I would replace this with a `cudaMemsetAsync()` on `stream`.  Since it is only setting to 0, you can use a memset.